bylaws/008-psr-evolution.md

```diff
@@ -0,0 +1,69 @@
+# PSR evolution
+
+## Scope and objectives
+
+A PSR is often comprised of text and code, more specifically interfaces. Those interfaces are pieces of code that are released and tagged at a specific moment in time, but the PHP language doesn't stand still... it evolves over time.
+
+This means that those interfaces need to withstand that evolution, and sometimes they need to be updated, to leverage new language features that could help better enforce the behaviors exposed in the PSR itself.
+At the same time, a PSR cannot be changed after its release (only erratas allowed), to protect a package that declared compatibility from becoming de-facto not compatible anymore.
+
+This document defines a process to be followed in updating PSR interfaces, in a way that is not breaking in regard to behavior for end users, and with an appropriate upgrade path for the packages.
+
+## New releases
+
+A new minor release of a PHP-FIG package containing interfaces for a PSR MUST follow these rules:
+ * the new release MUST follow [Semantic Versioning](https://semver.org/) rules;
+ * the PSR behavior MUST NOT be altered;
+ * the packages that implement the interfaces or users that are consuming them MUST NOT suffer breaking changes;
+ * the PHP version constraint of the PHP-FIG package MAY be altered to require newer language features that would aid cross-compatibility;
+ * the PHP version constraint of the PHP-FIG package MUST NOT be altered to use newer language features that would create cross-compatibility issues.
+ 
+A new major release of a PHP-FIG package containing interfaces for a PSR MUST follow the same rules, with this exception:
+ * the new major version of the package MAY contain breaking changes if the implementing packages have a reasonable upgrade path, like the possibility of releasing a cross-compatible implementation with the previous releases;
+ * the new major version of the package MAY refer to a new, superseding PSR.
+
+### Workflow
+
+Since releasing new versions of the interfaces MUST NOT alter the PSR in its behavior, those releases can be voted in with the same process as errata changes. The new releases MUST be declared and embedded in the PSR document, with eventual indications on the upgrade path in the meta document.
```

Comment:
This would be the first time we make non-trivial changes to the spec document rather than the metadoc.  It... kinda makes sense in context here but the end result is that PSR-3 would end up with 3 versions of all its code in the spec doc.  That's... a lot of change.

I don't have a better idea off hand, but this worries me.

Reply:
It depends on what is meant by "declared and embedded in the PSR document". Does this mean embedding the different versions of the interface? Or does it mean a link to the tagged version of the new interfaces in their repo (e.g., https://github.com/php-fig/http-message/releases/tag/1.0.1)?

We need clarification on what that phrase means.

Reply:
I had in mind adding a new code snippet with the changes, something along the lines of "Since version 1.1, the interface was changed to receive argument types: <code snippet>".

Would this phrasing be fine for you?

```suggestion
Since releasing new versions of the interfaces MUST NOT alter the PSR in its behavior, those releases can be voted in with the same process as errata changes. The new releases MUST be declared and embedded in the PSR document following the original code snippets in the original document, with eventual indications on the upgrade path in the meta document.
```

Reply:
The problem with saying "following the original code snippets" is that the current "code snippets" are full interface definitions generally - or do you mean something like this:

```markdown
\`\`\`php
interface ContainerInterface
{
    /**
     * Test whether or not a service is registered in the container.
     *
     * @param string $name
     * @return bool
     */
    public function has($name);


    /**
     * Retrieve a service from the container:
     *
     * @param string $name
     * @return object
     */
    public function get($name);
}
\`\`\`

> Since psr/container version 1.1, the above interface has been updated to add argument type hints.
```

(added escape characters on the embedded code sample so it would highlight properly)

Basically, I think even with clarifications to the verbiage, we may also need an example.

Reply:
Is e0c9954 enough for this?

Reply:
Maybe there is another way to do this entirely: we are already versioning the spec document with Git, so why don't we just update the document with the latest changes, specifying the version in the header, and list all the various versions in the website?

Reply:
I fear that that approach would complicate the site a lot, since it would have to use the submodule at multiple commits, and commits would've have to be fixed. Or we would have to have branches. Too complicated for a simple document.